File: src/routing/secure-routing.cpp
#include "secure-routing.h"
#include <cmath>
#include <algorithm>
#include <stdexcept>
#include <sstream>

namespace vanet {
namespace routing {

// Constants for protocol parameters
constexpr double MAX_TRUST_SCORE = 1.0;
constexpr double MIN_TRUST_SCORE = 0.0;
constexpr double TRUST_THRESHOLD = 0.5;
constexpr double MAX_SPEED = 200.0; // km/h
constexpr double MAX_ACCELERATION = 10.0; // m/s^2
constexpr std::chrono::seconds ROUTE_TIMEOUT(60);
constexpr std::chrono::seconds NEIGHBOR_TIMEOUT(10);
constexpr uint32_t MAX_HOP_COUNT = 10;

SecureRoutingProtocol::SecureRoutingProtocol(const std::string& id) 
    : vehicleId(id), cryptoModule(std::make_unique<crypto::CryptoModule>()) {
    localInfo.id = id;
    localInfo.trustScore = MAX_TRUST_SCORE;
}

SecureRoutingProtocol::~SecureRoutingProtocol() = default;

bool SecureRoutingProtocol::initializeVehicle(const VehicleInfo& info) {
    if (info.id != vehicleId) {
        return false;
    }
    
    localInfo = info;
    return cryptoModule->generateKeyPair();
}

bool SecureRoutingProtocol::updatePosition(const Position& newPos) {
    if (!isValidMovement(localInfo.position, newPos, 
        std::chrono::duration_cast<std::chrono::seconds>(
            newPos.timestamp - localInfo.position.timestamp).count())) {
        return false;
    }
    
    localInfo.position = newPos;
    pruneExpiredEntries();
    return true;
}

bool SecureRoutingProtocol::sendData(const std::string& destination, const std::vector<uint8_t>& data) {
    auto routeIt = routingTable.find(destination);
    if (routeIt == routingTable.end()) {
        if (!findRoute(destination)) {
            return false;
        }
        routeIt = routingTable.find(destination);
    }
    
    if (!isVehicleTrusted(routeIt->second.nextHop)) {
        invalidateRoute(destination);
        return false;
    }
    
    // Create secure message
    auto message = createRoutingMessage(MessageType::DATA, destination);
    message.insert(message.end(), data.begin(), data.end());
    
    // Sign and encrypt using crypto module
    auto secureMessage = cryptoModule->createSecureMessage(message);
    
    // Send to next hop (implementation depends on network layer)
    // For simulation purposes, this would interface with NS-3
    return true;
}

bool SecureRoutingProtocol::receiveMessage(const std::vector<uint8_t>& message) {
    if (!verifyRoutingMessage(message)) {
        return false;
    }
    
    // Extract message type and handle accordingly
    MessageType type = static_cast<MessageType>(message[0]);
    switch (type) {
        case MessageType::HELLO:
            return processBeacon(message);
        case MessageType::ROUTE_REQUEST:
            // Handle route request
            break;
        case MessageType::ROUTE_REPLY:
            // Handle route reply
            break;
        case MessageType::ROUTE_ERROR:
            // Handle route error
            break;
        case MessageType::DATA:
            // Handle data message
            break;
    }
    
    return true;
}

bool SecureRoutingProtocol::findRoute(const std::string& destination) {
    // Create RREQ message
    auto rreq = createRoutingMessage(MessageType::ROUTE_REQUEST, destination);
    
    // Sign message
    auto secureRreq = cryptoModule->createSecureMessage(rreq);
    
    // Broadcast to neighbors (implementation depends on network layer)
    // For simulation purposes, this would interface with NS-3
    
    return true;
}

bool SecureRoutingProtocol::updateRoute(const std::string& destination, const RouteEntry& entry) {
    if (entry.hopCount >= MAX_HOP_COUNT) {
        return false;
    }
    
    auto now = std::chrono::system_clock::now();
    if (entry.timestamp + ROUTE_TIMEOUT < now) {
        return false;
    }
    
    routingTable[destination] = entry;
    return true;
}

bool SecureRoutingProtocol::invalidateRoute(const std::string& destination) {
    auto it = routingTable.find(destination);
    if (it != routingTable.end()) {
        routingTable.erase(it);
        
        // Create and broadcast RERR message
        auto rerr = createRoutingMessage(MessageType::ROUTE_ERROR, destination);
        auto secureRerr = cryptoModule->createSecureMessage(rerr);
        
        // Broadcast RERR (implementation depends on network layer)
        return true;
    }
    return false;
}

double SecureRoutingProtocol::calculateTrust(const std::string& vehicleId) {
    auto it = trustScores.find(vehicleId);
    if (it == trustScores.end()) {
        return MIN_TRUST_SCORE;
    }
    
    // Factor in various trust metrics
    double score = it->second;
    
    // Check for suspicious behavior
    if (detectBlackHole(vehicleId) || detectSybil(vehicleId)) {
        score *= 0.5;
    }
    
    // Verify position consistency
    auto neighborIt = neighborTable.find(vehicleId);
    if (neighborIt != neighborTable.end()) {
        if (detectPositionFalsification(vehicleId, neighborIt->second.position)) {
            score *= 0.5;
        }
    }
    
    return std::clamp(score, MIN_TRUST_SCORE, MAX_TRUST_SCORE);
}

void SecureRoutingProtocol::updateTrustScore(const std::string& vehicleId, double score) {
    double currentScore = trustScores[vehicleId];
    // Use exponential moving average
    constexpr double alpha = 0.3;
    trustScores[vehicleId] = (alpha * score) + ((1 - alpha) * currentScore);
}

bool SecureRoutingProtocol::isVehicleTrusted(const std::string& vehicleId) {
    return calculateTrust(vehicleId) >= TRUST_THRESHOLD;
}

void SecureRoutingProtocol::sendBeacon() {
    auto beacon = createRoutingMessage(MessageType::HELLO, "");
    auto secureBeacon = cryptoModule->createSecureMessage(beacon);
    
    // Broadcast beacon (implementation depends on network layer)
}

bool SecureRoutingProtocol::processBeacon(const std::vector<uint8_t>& beacon) {
    // Extract vehicle info from beacon
    VehicleInfo info;
    // ... parse beacon data ...
    
    // Verify certificate
    if (!cryptoModule->verifySecureMessage(crypto::CryptoModule::SecureMessage{/*...*/})) {
        return false;
    }
    
    // Update neighbor table
    neighborTable[info.id] = info;
    
    // Update trust score based on beacon validity
    updateTrustScore(info.id, 1.0);
    
    return true;
}

bool SecureRoutingProtocol::detectBlackHole(const std::string& suspectId) {
    // Check for abnormally high route advertisements
    // and low packet forwarding rates
    return false;
}

bool SecureRoutingProtocol::detectSybil(const std::string& suspectId) {
    // Check for multiple identities from similar positions
    // or with similar certificates
    return false;
}

bool SecureRoutingProtocol::detectReplay(const std::vector<uint8_t>& message) {
    return cryptoModule->isReplayMessage(crypto::CryptoModule::SecureMessage{/*...*/});
}

bool SecureRoutingProtocol::detectPositionFalsification(const std::string& vehicleId, const Position& reportedPos) {
    auto it = neighborTable.find(vehicleId);
    if (it == neighborTable.end()) {
        return false;
    }
    
    const auto& lastPos = it->second.position;
    double timeElapsed = std::chrono::duration_cast<std::chrono::seconds>(
        reportedPos.timestamp - lastPos.timestamp).count();
    
    return !isValidMovement(lastPos, reportedPos, timeElapsed);
}

std::vector<uint8_t> SecureRoutingProtocol::createRoutingMessage(MessageType type, const std::string& destination) {
    std::vector<uint8_t> message;
    message.push_back(static_cast<uint8_t>(type));
    
    // Add source ID
    message.insert(message.end(), vehicleId.begin(), vehicleId.end());
    message.push_back(0); // null terminator
    
    // Add destination ID
    message.insert(message.end(), destination.begin(), destination.end());
    message.push_back(0); // null terminator
    
    // Add timestamp
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()).count();
    message.insert(message.end(),
        reinterpret_cast<uint8_t*>(&timestamp),
        reinterpret_cast<uint8_t*>(&timestamp) + sizeof(timestamp));
    
    return message;
}

bool SecureRoutingProtocol::verifyRoutingMessage(const std::vector<uint8_t>& message) {
    if (message.empty()) {
        return false;
    }
    
    // Verify message structure
    MessageType type = static_cast<MessageType>(message[0]);
    if (type > MessageType::DATA) {
        return false;
    }
    
    // Verify signature and certificate
    return cryptoModule->verifySecureMessage(crypto::CryptoModule::SecureMessage{/*...*/});
}

double SecureRoutingProtocol::calculateDistance(const Position& pos1, const Position& pos2) {
    double dx = pos1.x - pos2.x;
    double dy = pos1.y - pos2.y;
    double dz = pos1.z - pos2.z;
    return std::sqrt(dx*dx + dy*dy + dz*dz);
}

bool SecureRoutingProtocol::isValidMovement(const Position& oldPos, const Position& newPos, double timeElapsed) {
    if (timeElapsed <= 0) {
        return false;
    }
    
    double distance = calculateDistance(oldPos, newPos);
    double speed = distance / timeElapsed;
    
    // Convert speed to km/h
    speed = speed * 3.6;
    
    // Check if speed is within reasonable limits
    if (speed > MAX_SPEED) {
        return false;
    }
    
    // Check acceleration
    double acceleration = speed / timeElapsed;
    return acceleration <= MAX_ACCELERATION;
}

void SecureRoutingProtocol::pruneExpiredEntries() {
    auto now = std::chrono::system_clock::now();
    
    // Prune routing table
    for (auto it = routingTable.begin(); it != routingTable.end();) {
        if (it->second.timestamp + ROUTE_TIMEOUT < now) {
            it = routingTable.erase(it);
        } else {
            ++it;
        }
    }
    
    // Prune neighbor table
    for (auto it = neighborTable.begin(); it != neighborTable.end();) {
        if (it->second.position.timestamp + NEIGHBOR_TIMEOUT < now) {
            it = neighborTable.erase(it);
        } else {
            ++it;
        }
    }
}

} // namespace routing
} // namespace vanet 

File: src/routing/secure_routing.py
from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple
import time
import math
from ..crypto.crypto_module import CryptoModule, SecureMessage

@dataclass
class Position:
    x: float
    y: float
    z: float
    timestamp: float

@dataclass
class VehicleInfo:
    id: str
    position: Position
    speed: float
    direction: float
    trust_score: float
    certificate: Optional[bytes] = None

@dataclass
class RouteEntry:
    next_hop: str
    hop_count: int
    timestamp: float
    trust_score: float

class MessageType:
    HELLO = 0
    ROUTE_REQUEST = 1
    ROUTE_REPLY = 2
    ROUTE_ERROR = 3
    DATA = 4

class SecureRoutingProtocol:
    def __init__(self, vehicle_id: str):
        self.vehicle_id = vehicle_id
        self.crypto_module = CryptoModule()
        self.local_info = None
        self.routing_table: Dict[str, RouteEntry] = {}
        self.neighbor_table: Dict[str, VehicleInfo] = {}
        self.trust_scores: Dict[str, float] = {}
        self.message_tracking: Dict[str, Tuple[int, float]] = {}  # (last_sequence, last_update)
        
        # Constants
        self.MAX_TRUST_SCORE = 1.0
        self.MIN_TRUST_SCORE = 0.0
        self.TRUST_THRESHOLD = 0.5
        self.MAX_SPEED = 200.0  # km/h
        self.MAX_ACCELERATION = 10.0  # m/s²
        self.ROUTE_TIMEOUT = 60  # seconds
        self.NEIGHBOR_TIMEOUT = 10  # seconds
        self.MAX_HOP_COUNT = 10
        
        # Initialize cryptographic keys
        self.crypto_module.generate_key_pair()

    def initialize_vehicle(self, info: VehicleInfo) -> bool:
        """Initialize the vehicle with its information."""
        if info.id != self.vehicle_id:
            return False
        
        self.local_info = info
        return True

    def update_position(self, new_pos: Position) -> bool:
        """Update vehicle's position if the movement is valid."""
        if not self.local_info:
            return False
            
        if not self._is_valid_movement(self.local_info.position, new_pos):
            return False
            
        self.local_info.position = new_pos
        self._prune_expired_entries()
        return True

    def send_data(self, destination: str, data: bytes) -> bool:
        """Send data to a destination using secure routing."""
        # Check if we have a route
        route = self.routing_table.get(destination)
        if not route:
            if not self.find_route(destination):
                return False
            route = self.routing_table.get(destination)
            
        if not route or not self.is_vehicle_trusted(route.next_hop):
            return False
            
        # Create secure message
        message = self._create_routing_message(MessageType.DATA, destination, data)
        secure_message = self.crypto_module.create_secure_message(message)
        
        # In a real implementation, this would send the message to the next hop
        # For simulation, we'll just return True
        return True

    def receive_message(self, message: bytes) -> bool:
        """Process received message."""
        try:
            secure_message = self._deserialize_message(message)
            if not self.crypto_module.verify_secure_message(secure_message):
                return False
                
            message_type = message[0]
            if message_type == MessageType.HELLO:
                return self.process_beacon(secure_message.payload)
            elif message_type == MessageType.ROUTE_REQUEST:
                # Handle route request
                pass
            elif message_type == MessageType.ROUTE_REPLY:
                # Handle route reply
                pass
            elif message_type == MessageType.ROUTE_ERROR:
                # Handle route error
                pass
            elif message_type == MessageType.DATA:
                # Handle data message
                pass
                
            return True
        except Exception as e:
            print(f"Error processing message: {e}")
            return False

    def find_route(self, destination: str) -> bool:
        """Initiate route discovery to destination."""
        message = self._create_routing_message(MessageType.ROUTE_REQUEST, destination)
        secure_message = self.crypto_module.create_secure_message(message)
        
        # In a real implementation, this would broadcast the route request
        # For simulation, we'll just return True
        return True

    def update_route(self, destination: str, entry: RouteEntry) -> bool:
        """Update routing table entry."""
        if entry.hop_count >= self.MAX_HOP_COUNT:
            return False
            
        if time.time() - entry.timestamp > self.ROUTE_TIMEOUT:
            return False
            
        self.routing_table[destination] = entry
        return True

    def calculate_trust(self, vehicle_id: str) -> float:
        """Calculate trust score for a vehicle."""
        if vehicle_id not in self.trust_scores:
            return self.MIN_TRUST_SCORE
            
        score = self.trust_scores[vehicle_id]
        
        # Check for suspicious behavior
        if self.detect_black_hole(vehicle_id) or self.detect_sybil(vehicle_id):
            score *= 0.5
            
        # Verify position consistency
        if vehicle_id in self.neighbor_table:
            if self.detect_position_falsification(vehicle_id, self.neighbor_table[vehicle_id].position):
                score *= 0.5
                
        return max(min(score, self.MAX_TRUST_SCORE), self.MIN_TRUST_SCORE)

    def update_trust_score(self, vehicle_id: str, score: float):
        """Update trust score using exponential moving average."""
        current_score = self.trust_scores.get(vehicle_id, self.MAX_TRUST_SCORE)
        alpha = 0.3  # Weight for new score
        self.trust_scores[vehicle_id] = (alpha * score) + ((1 - alpha) * current_score)

    def is_vehicle_trusted(self, vehicle_id: str) -> bool:
        """Check if a vehicle's trust score is above threshold."""
        return self.calculate_trust(vehicle_id) >= self.TRUST_THRESHOLD

    def send_beacon(self):
        """Send periodic beacon message."""
        message = self._create_routing_message(MessageType.HELLO, "")
        secure_message = self.crypto_module.create_secure_message(message)
        
        # In a real implementation, this would broadcast the beacon
        # For simulation, we'll just pass

    def process_beacon(self, beacon: bytes) -> bool:
        """Process received beacon message."""
        try:
            info = self._deserialize_vehicle_info(beacon)
            if not info:
                return False
                
            self.neighbor_table[info.id] = info
            self.update_trust_score(info.id, 1.0)
            return True
        except Exception:
            return False

    def detect_black_hole(self, suspect_id: str) -> bool:
        """Detect black hole attack behavior."""
        # Implementation would track packet forwarding rates
        # and route advertisement patterns
        return False

    def detect_sybil(self, suspect_id: str) -> bool:
        """Detect Sybil attack behavior."""
        # Implementation would check for multiple identities
        # from similar positions
        return False

    def detect_position_falsification(self, vehicle_id: str, reported_pos: Position) -> bool:
        """Detect position falsification attack."""
        if vehicle_id not in self.neighbor_table:
            return False
            
        last_pos = self.neighbor_table[vehicle_id].position
        time_elapsed = reported_pos.timestamp - last_pos.timestamp
        
        return not self._is_valid_movement(last_pos, reported_pos)

    def _create_routing_message(self, msg_type: int, destination: str, data: bytes = b"") -> bytes:
        """Create a routing message."""
        message = bytearray([msg_type])
        message.extend(self.vehicle_id.encode())
        message.append(0)  # null terminator
        message.extend(destination.encode())
        message.append(0)  # null terminator
        message.extend(str(time.time()).encode())
        if data:
            message.extend(data)
        return bytes(message)

    def _deserialize_message(self, message: bytes) -> SecureMessage:
        """Deserialize a received message."""
        # Implementation depends on your message format
        # This is a simplified example
        return SecureMessage(
            payload=message,
            signature=b"",
            timestamp=time.time(),
            sequence_number=0
        )

    def _deserialize_vehicle_info(self, data: bytes) -> Optional[VehicleInfo]:
        """Deserialize vehicle information from bytes."""
        # Implementation depends on your message format
        # This is a simplified example
        return None

    def _is_valid_movement(self, old_pos: Position, new_pos: Position) -> bool:
        """Check if movement between positions is physically possible."""
        if not old_pos or not new_pos:
            return True
            
        time_elapsed = new_pos.timestamp - old_pos.timestamp
        if time_elapsed <= 0:
            return False
            
        distance = self._calculate_distance(old_pos, new_pos)
        speed = distance / time_elapsed
        
        # Convert speed to km/h
        speed = speed * 3.6
        
        # Check if speed is within reasonable limits
        if speed > self.MAX_SPEED:
            return False
            
        # Check acceleration
        acceleration = speed / time_elapsed
        return acceleration <= self.MAX_ACCELERATION

    def _calculate_distance(self, pos1: Position, pos2: Position) -> float:
        """Calculate Euclidean distance between two positions."""
        return math.sqrt(
            (pos1.x - pos2.x) ** 2 +
            (pos1.y - pos2.y) ** 2 +
            (pos1.z - pos2.z) ** 2
        )

    def _prune_expired_entries(self):
        """Remove expired entries from routing and neighbor tables."""
        current_time = time.time()
        
        # Prune routing table
        expired_routes = [
            dest for dest, entry in self.routing_table.items()
            if current_time - entry.timestamp > self.ROUTE_TIMEOUT
        ]
        for dest in expired_routes:
            del self.routing_table[dest]
            
        # Prune neighbor table
        expired_neighbors = [
            vid for vid, info in self.neighbor_table.items()
            if current_time - info.position.timestamp > self.NEIGHBOR_TIMEOUT
        ]
        for vid in expired_neighbors:
            del self.neighbor_table[vid] 